tests/test_scan.py

```diff
@@ -0,0 +1,518 @@
+    # Now, we rerun the test, but use the "replace" features of
+    # `convert_outer_out_to_in`
+    #
+    test_scan_args, new_oi_var = convert_outer_out_to_in(
+        input_scan_args,
+        var,
+        inner_out_fn=create_inner_out_logp,
+        output_scan_args=input_scan_args,
+    )
+
+    scan_out = construct_scan(test_scan_args)
+
+    #
+    # Evaluate the manually and automatically constructed log-likelihoods and
+    # compare.
+    #
+    res = scan_out[var_idx].eval({new_oi_var: Y_obs.value})
+    exp_res = Y_logp.eval()
+
+    assert np.array_equal(res, exp_res)
+
+
+def test_convert_outer_out_to_in_hmm():
+    hmm_model_env = create_test_hmm()
+    input_scan_args = hmm_model_env["scan_args"]
+    M_tt = hmm_model_env["M_tt"]
+    N_tt = hmm_model_env["N_tt"]
+    mus_tt = hmm_model_env["mus_tt"]
+    sigmas_tt = hmm_model_env["sigmas_tt"]
+    Y_rv = hmm_model_env["Y_rv"]
+    S_0_rv = hmm_model_env["S_0_rv"]
+    Gamma_rv = hmm_model_env["Gamma_rv"]
+    rng_tt = hmm_model_env["rng"]
+
+    test_point = {
+        M_tt: 2,
+        N_tt: 10,
+        mus_tt: mus_tt.tag.test_value,
+    }
+    Y_obs = at.as_tensor_variable(Y_rv.eval(test_point))
+    S_obs = at.as_tensor_variable(S_0_rv.eval(test_point))
+    Y_obs.name = "Y_obs"
+
+    def logp_scan_fn(y_t, mus_t, sigma_t, S_tm1, Gamma_t, rng):
+        gamma_t = Gamma_t[S_tm1]
+        gamma_t.name = "gamma_t"
+        S_t = at.random.categorical(gamma_t, rng=rng, name="S_t")
+        mu_t = mus_t[S_t]
+        mu_t.name = "mu[S_t]"
+        Y_logp_t = logprob(at.random.normal(mu_t, sigma_t), y_t)
+        Y_logp_t.name = "logp(y_t)"
+        return S_t, Y_logp_t
+
+    (S_rv, Y_logp), scan_updates = aesara.scan(
+        fn=logp_scan_fn,
+        sequences=[Y_obs, mus_tt, sigmas_tt],
+        non_sequences=[Gamma_rv, rng_tt],
+        outputs_info=[{"initial": S_0_rv, "taps": [-1]}, {}],
+        strict=True,
+        name="scan_rv",
+    )
+    Y_logp.name = "Y_logp"
+
+    var_idx, var, io_var = get_random_outer_outputs(input_scan_args)[1]
+    value_map = {Y_rv: Y_obs, S_rv: S_obs}
+    test_scan_args, new_oi_var = convert_outer_out_to_in(
+        input_scan_args,
+        [var],
+        value_map,
+        inner_out_fn=create_inner_out_logp,
+        output_scan_args=input_scan_args,
+    )
+
+    scan_out = construct_scan(test_scan_args)
+    test_Y_logp = scan_out[var_idx]
+
+    #
+    # Evaluate the manually and automatically constructed log-likelihoods and
+    # compare.
+    #
+    new_test_point = dict(test_point)
+    new_test_point[new_oi_var] = Y_obs.value
+
+    # We need to reset the RNG each time, because `S_t` is still a
+    # `RandomVariable`
+    with aesara.change_flags(on_unused_input="warn"):
+        res = test_Y_logp.eval(new_test_point)
+
+    exp_res = Y_logp.eval(test_point)
+
+    assert np.array_equal(res, exp_res)
+
+
+def test_scan_logprob():
+    # srng = at.random.RandomStream()
+    rng = aesara.shared(np.random.default_rng(1234))
```

Comment:
Temporarily changed `RandomStreams` to shared `Generators` for the current development run/testing purposes.

Reply:
No problem; I don't think there's a strict requirement for either one, but both should be expected to work (although the `assert`s may need to change in response).